### PR TITLE
[AMDGPU][GlobalISel] Align `selectVOP3PMadMixModsImpl` with the `SelectionDAG` counterpart

### PR DESCRIPTION
**Pull request by shiltian**
The current `selectVOP3PMadMixModsImpl` can produce `V_MAD_FIX_F32` instruction
that violates constant bus restriction, while its `SelectionDAG` counterpart
doesn't. The culprit is in the copy stripping while the `SelectionDAG` version
only has a bitcast stripping. This PR simply aligns the two version.

### CODE REVIEW

**Comment by shiltian** [comment]
* **#110168** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/110168?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by shiltian** [comment]
I understand that this seems to revert some optimization, so I'm happy to hear about better approaches.

**Comment by petar-avramovic** [comment]
IMO, the easiest approach is to force vgpr reg class on all operands when selecting vop3p instructions. Do this with constrainOperandRegClass, to vgpr reg class you can get from getVGPRClassForBitWidth. This will insert copy from sgpr to vgpr for all operands.
SIFoldOperands can eliminate some of the copies by replacing vgpr with sgpr so you end up with same result in the end (SIFoldOperands knows about constant bus restriction).

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:5328]: This is just `Mods ^= ModsTmp` now?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir:1]: Is this test case `llvm-reduce`d already?

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir:1]: Now it is. :-)

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
Comment (shiltian) [reply llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:5328]: Not really. That is only the case for `NEG`.

**Comment by shiltian** [comment]
> IMO, the easiest approach is to force vgpr reg class on all operands when selecting vop3p instructions.

IIUC, this PR made a right move by removing the copy stripping.

**Comment by shiltian** [comment]
I added the following code into `selectVOP3PMadMixModsExt` and `selectVOP3PMadMixMods`. Not sure if this is what you suggested, but it's not able to fix the issue.

```
diff --git a/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp b/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
index febf0711c7d5..d95e0ee944bf 100644
--- a/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
+++ b/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
@@ -5366,6 +5366,13 @@ AMDGPUInstructionSelector::selectVOP3PMadMixModsExt(
   if (!Matched)
     return {};
 
+  const TargetRegisterClass *SrcRC =
+      TRI.getConstrainedRegClassForOperand(Root, *MRI);
+  const TargetRegisterClass *VGPRSrc =
+      TRI.getVGPRClassForBitWidth(AMDGPU::getRegBitWidth(SrcRC->getID()));
+  MachineInstr *I = Root.getParent();
+  Src = constrainOperandRegClass(*MF, TRI, *MRI, TII, RBI, *I, *VGPRSrc, Root);
+
   return {{
       [=](MachineInstrBuilder &MIB) { MIB.addReg(Src); },
       [=](MachineInstrBuilder &MIB) { MIB.addImm(Mods); } // src_mods
@@ -5379,6 +5386,13 @@ AMDGPUInstructionSelector::selectVOP3PMadMixMods(MachineOperand &Root) const {
   bool Matched;
   std::tie(Src, Mods) = selectVOP3PMadMixModsImpl(Root, Matched);
 
+  const TargetRegisterClass *SrcRC =
+      TRI.getConstrainedRegClassForOperand(Root, *MRI);
+  const TargetRegisterClass *VGPRSrc =
+      TRI.getVGPRClassForBitWidth(AMDGPU::getRegBitWidth(SrcRC->getID()));
+  MachineInstr *I = Root.getParent();
+  Src = constrainOperandRegClass(*MF, TRI, *MRI, TII, RBI, *I, *VGPRSrc, Root);
+
   return {{
       [=](MachineInstrBuilder &MIB) { MIB.addReg(Src); },
       [=](MachineInstrBuilder &MIB) { MIB.addImm(Mods); } // src_mods
```

**Comment by shiltian** [comment]
Overall we can make it align with the `SelectionDAG` counterpart, unblock the PR depending on it, and then improve it using what @petar-avramovic suggested on both side.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir:41]: It's preferable to have IR tests for this issue. This also should test the FMA case

File: llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/madmix-constant-bus-violation.mir:40]: The rcp isn't relevant for this issue

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:5315]: I think we need new tests where a bitcast is stripped

**Review by arsenm — COMMENTED** [review]
I'm not sure I trust this. It might be better to directly check the constant bus restriction, similar to how ThreeOpFrag does

**Comment by petar-avramovic** [comment]
Hi @shiltian, I did not look into the patch carefully, can you provide LLVM IR test that violates constant bus restriction first.

What I had in mind was that you somehow end up with sgpr register after selectVOP3PMadMixModsImpl

If you want to use existing functions you could
- add it to machine instruction **first** (without checking anything).
- **then** constrain its register class to vgpr.
expected result is that if sgpr was added to instruction, constrainOperandRegClass(...,*VGPRSrc, SrcOp) would replace it by inserting a copy to vgpr.
The constrainSelectedInstRegOperands allows both sgpr and vgpr for that operand. We have to explicitly constrain all operands to vgpr. Other option is to do this in tablegen. End result should be the same.

```
diff --git a/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp b/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
index febf0711c7d5..9a81e7609dac 100644
--- a/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
+++ b/llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp
@@ -5356,6 +5356,13 @@ AMDGPUInstructionSelector::selectVOP3PMadMixModsImpl(MachineOperand &Root,
   return {Src, Mods};
 }
 
+MachineOperand &getLastOp(MachineInstrBuilder &MIB) {
+  unsigned Idx = MIB->getNumOperands() - 1;
+  while (MIB->getOperand(Idx).isImplicit())
+    --Idx;
+  return MIB->getOperand(Idx);
+}
+
 InstructionSelector::ComplexRendererFns
 AMDGPUInstructionSelector::selectVOP3PMadMixModsExt(
     MachineOperand &Root) const {
@@ -5367,7 +5374,14 @@ AMDGPUInstructionSelector::selectVOP3PMadMixModsExt(
     return {};
 
   return {{
-      [=](MachineInstrBuilder &MIB) { MIB.addReg(Src); },
+      [=](MachineInstrBuilder &MIB) {
+        MIB.addReg(Src);
+        const TargetRegisterClass *VGPRSrc = &AMDGPU::VGPR_32RegClass;
+        MachineOperand &SrcOp = getLastOp(MIB);
+        constrainOperandRegClass(*MF, TRI, *MRI, TII, RBI, *MIB, *VGPRSrc,
+                                 SrcOp);
+      },
       [=](MachineInstrBuilder &MIB) { MIB.addImm(Mods); } // src_mods
   }};
 }
```
I added snippet that I would assume works, but don't have test for it.

**Comment by shiltian** [comment]
> I'm not sure I trust this.

Which part? https://github.com/llvm/llvm-project/blob/83fe85115da9dc25fa270d2ea8140113c8d49670/llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp#L3481

> It might be better to directly check the constant bus restriction, similar to how ThreeOpFrag does

I tried this as you suggested but because the pattern here is a `ComplexPattern` (or `GIComplexOperandMatcher`), it doesn't have support for checking predicate, `GISelPredicateCode` as what `ThreeOpFrag` is used.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir:2]: Don't need -global-isel flag when running on mir tests

**Review by arsenm — COMMENTED** [review]
Should have end to end IR test, including bitcasted cases

**Comment by shiltian** [comment]
> Should have end to end IR test

How to make a IR test using `mad`? LLVM has `fma` but no `mad`. This sequence of code was reduced from the MIR sequence generated in #109295 but that PR is blocked by this one.

**Comment by shiltian** [comment]
Thanks @petar-avramovic . It indeed works. It might be better to do it in TableGen so I will probably make a follow-up patch. For now, I'd like to get the issue resolved such that another PR can be merged as well.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll:77]: This regressed gfx10 which has the more relaxed constant bus limit

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll:4]: Don't need amdgpu_ps Should use better test name

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll:2]: Don't need -verify-machininstrs. Also prefer to just use the explicit denormal-fp-math-f32 attribute in the test itself

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.ll:23]: Test all the combinations of SGPR in each position. We probably already have a comprehensive test for selectiondag, maybe should just add a run line there

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll:77]: The `*v_fma_mix_f32` instruction still uses two SGPRs. The only issue here is, it can't see "through" the copies such that it can't directly use the source operands of the shifts via `op_sel`. I think the ISel version also has this limitation (apparently because here we "copied" the logic).

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir:2]: Since we have end-to-end IR tests, can I just remove this two MIR tests?

**Comment by shiltian** [comment]
bump

File: llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/combine-fma-add-ext-mul.ll:77]: Still should fix this in a follow up

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/fmamix-constant-bus-violation.mir:2]: Should keep both probably

**Review by arsenm — APPROVED** [review]

**Pull request merged**
